design/ceph/object/swift-and-keystone-integration.md

```diff
@@ -0,0 +1,323 @@
+---
+target-version: release-1.9
+---
+
+# Swift and Keystone Integration
+
+## Summary
+
+### Goals
+
+The goal of this proposal is to allow configuring the Swift API and
+the Keystone integration of Ceph RGW natively via the Object Store
+CRD, which allows native integration of the Rook operated Ceph RGW into
+OpenStack clouds.
+
+Both changes are bundled together as one proposal because they will
+typically deployed together. It is unlikely to encounter a Swift
+object store API outside an OpenStack cloud and in the context of an
+OpenStack cloud the Swift API needs to be integrated with the Keystone
+authentication service.
+
+The Keystone integration must support the current [OpenStack Identity
+API version 3](https://docs.openstack.org/api-ref/identity/v3/).
+
+It must be possible to serve S3 and Swift for the same object store
+pool.
+
+It must be possible to [obtain S3 credentials via OpenStack](
+https://docs.ceph.com/en/octopus/radosgw/keystone/#keystone-integration-with-the-s3-api).
+
+Any changes to the CRD must be future safe and cleanly allow extension
+to further technologies (such as LDAP authentication).
+
+### Non-Goals
+
+* Support for OpenStack Identity API versions below v3. API version v2 has long
+  been deprecated and [has been removed in the "queens" version of
+  Keystone](https://docs.openstack.org/keystone/xena/contributor/http-api.html)
+  which was released in 2018 and is now in extended maintenance mode
+  (which means it gets no more points releases and only sporadic bug
+  fixes/security fixes).
+
+* Authenticating Ceph RGW to Keystone via admin token (a.k.a. shared secret).
+  This is a deliberate choice as [admin tokens should not be used in production environments](
+  https://docs.openstack.org/keystone/rocky/admin/identity-bootstrap.html#using-a-shared-secret).
+
+* Support for APIs beside S3 and Swift.
+
+* Interaction of Swift with OBCs is out of scope of this document.  If
+  you need to access a bucket created by OBC via Swift you need to
+  create a separate `cephobjectstoreuser`, configure its access rights
+  to the bucket and use those credentials.
```

Comment:
@jklippel in the 2 years since this design was merged, we now also have COSI to consider -- it's the upcoming, eventual replacement for OBC. I think it's still reasonable to say that Swift-only CephObjectStores don't support OBCs or COSI, and I want to make sure that gets documented clearly for users. In #13807, Would you make sure to (1) update this design doc note, and (2) capture these limitations in user-facing docs well?

Reply:
https://github.com/rook/rook/pull/13807/commits/241bb57ab0f6ea8b6bf7b73afc7e75420f2024b8